pkg/agent/controller/networkpolicy/status_controller.go

```diff
@@ -0,0 +1,236 @@
+	c.ruleGenerationsLock.Lock()
+	defer c.ruleGenerationsLock.Unlock()
+	delete(c.ruleGenerations, ruleID)
+}
+
+func (c *StatusController) SetExpectation(policyID types.UID, generation int64) {
+	c.expectedGenerationsLock.Lock()
+	defer c.expectedGenerationsLock.Unlock()
+	c.expectedGenerations[policyID] = generation
+}
+
+func (c *StatusController) DeleteExpectation(policyID types.UID) {
+	c.expectedGenerationsLock.Lock()
+	defer c.expectedGenerationsLock.Unlock()
+	delete(c.expectedGenerations, policyID)
+}
+
+func (c *StatusController) Resync() {
+	klog.Info("Resyncing NetworkPolicy status")
+	c.expectedGenerationsLock.RLock()
+	defer c.expectedGenerationsLock.RUnlock()
+	for policy := range c.expectedGenerations {
+		c.queue.Add(policy)
+	}
+}
+
+// worker is a long-running function that will continually call the processNextWorkItem function in
+// order to read and process a message on the workqueue.
+func (c *StatusController) worker() {
+	for c.processNextWorkItem() {
+	}
+}
+
+func (c *StatusController) processNextWorkItem() bool {
+	obj, quit := c.queue.Get()
+	if quit {
+		return false
+	}
+	// We call Done here so the workqueue knows we have finished processing this item. We also
+	// must remember to call Forget if we do not want this work item being re-queued. For
+	// example, we do not call Forget if a transient error occurs, instead the item is put back
+	// on the workqueue and attempted again after a back-off period.
+	defer c.queue.Done(obj)
+
+	// We expect NetworkPolicy UID to come off the workqueue.
+	if key, ok := obj.(types.UID); !ok {
+		// As the item in the workqueue is actually invalid, we call Forget here else we'd
+		// go into a loop of attempting to process a work item that is invalid.
+		// This should not happen: enqueueNode only enqueues UIDs.
+		c.queue.Forget(obj)
+		klog.Errorf("Expected UID in work queue but got %#v", obj)
+		return true
+	} else if err := c.syncHandler(key); err == nil {
+		// If no error occurs we Forget this item so it does not get queued again until
+		// another change happens.
+		c.queue.Forget(key)
+	} else {
+		// Put the item back on the workqueue to handle any transient errors.
+		c.queue.AddRateLimited(key)
+		klog.Errorf("Error syncing NetworkPolicyStatus %s, requeuing. Error: %v", key, err)
+	}
+	return true
+}
+
+func (c *StatusController) syncHandler(uid types.UID) error {
+	c.expectedGenerationsLock.RLock()
+	expectedGeneration, exists := c.expectedGenerations[uid]
+	c.expectedGenerationsLock.RUnlock()
+	// No expectation on this Policy, no further processing.
+	if !exists {
+		return nil
+	}
+
+	rules := c.ruleCache.getRulesByNetworkPolicy(string(uid))
+	// The policy must have been deleted, no further processing.
+	if len(rules) == 0 {
+		return nil
+	}
+	if matched := func() bool {
+		c.ruleGenerationsLock.RLock()
+		defer c.ruleGenerationsLock.RUnlock()
+		for _, rule := range rules {
+			ruleGeneration, exists := c.ruleGenerations[rule.ID]
```

Comment:
Question - could we save rule generation in ruleCache, then we need not a separate map of ruleGenerations?

And could we maintain a counter of rules match the expected generation, then we need not to check all rules here?

Reply:
Yes, I meant a separate field for the realized generation in the rule struct.

> a rule can be realized multiple times if its associated groups are changed, we cannot just increase the counter when a rule is realized and need to know whether it has been realized before.
If we save the realized generation in the rule struct, we can check the current value of the field to decide whether or not to increment the counter?

Reply:
I realized the rule actually shouldn't be generation related because rules are immutable, otherwise changing rule B could cause rule A updated unnecessarily (if counting policy generation into the hash value). When we made a change to an existing rule, it's considered a new rule added and an old rule deleted. The policy should be considered realized when all of its current rules are realized and all of its removed rules are uninstalled. Changed the implementation in latest patch.